apps/web/src/components/forms/notification-form.tsx

```diff
@@ -46,39 +48,56 @@ interface Props {
   onSubmit?: () => void;
 }
 
+function getDefaultProviderData(defaultValues?: Notification) {
+  if (defaultValues?.provider === "email") {
+    return JSON.parse(defaultValues?.data).email;
+  } else if (defaultValues?.provider === "discord") {
+    return JSON.parse(defaultValues?.data).discord;
+  }
+}
+
 export function NotificationForm({
   workspaceSlug,
   defaultValues,
   onSubmit: onExternalSubmit,
 }: Props) {
   const [isPending, startTransition] = useTransition();
+  const [isTestPending, startTestTransition] = useTransition();
   const { toast } = useToastAction();
   const router = useRouter();
   const form = useForm<Notification>({
     resolver: zodResolver(insertNotificationSchema),
     defaultValues: {
       ...defaultValues,
-      data:
-        defaultValues?.provider === "email"
-          ? JSON.parse(defaultValues?.data).email
-          : "",
+      data: getDefaultProviderData(defaultValues),
     },
   });
+  const watchProvider = form.watch("provider");
 
   async function onSubmit({ provider, data, ...rest }: Notification) {
     startTransition(async () => {
       try {
+        let providerJSON = {};
+        switch (watchProvider) {
+          case "email":
+            providerJSON = { email: data };
+            break;
+          case "discord":
+            providerJSON = { discord: data };
+            break;
+        }
```

Comment:
Let's extract it also into a separate `function setProviderData()` as you did for the `getDefaultProviderData` and add the `default: return {}` case.